File: routes/compareProgress.js
const fs = require('fs');
const path = require('path');
const express = require('express');
const db = require('../db');
const { parseModelJson } = require('../lib/claudeJson');
const { resolveProvider } = require('../lib/ai/provider');
const { ok, fail } = require('../middleware/responses');

const router = express.Router();

// Shared with the desktop (Rust) build — keep prompt text in prompts/, not inline.
const SYSTEM_PROMPT = fs
  .readFileSync(path.join(__dirname, '../prompts/compare-progress.system.txt'), 'utf8')
  .trim();

function tagYesNo(v) {
  return v === true ? 'yes' : 'no';
}

function buildUserMessage(studentName, iepFlags, entries) {
  const n = entries.length;
  const iep = iepFlags != null && String(iepFlags).trim() ? String(iepFlags).trim() : 'None provided';
  const blocks = entries.map((e, i) => {
    const t = e.tags || {};
    return `--- Entry ${i + 1}: ${e.assignment_name} (${e.date}) ---
Word count: ${e.word_count ?? 0}
Tags: Claim: ${tagYesNo(t.claim_present)}, Evidence: ${tagYesNo(t.evidence_cited)}, Explanation: ${tagYesNo(t.explanation_present)}, Source Named: ${tagYesNo(t.source_named)}, Incomplete: ${tagYesNo(t.response_incomplete)}, AI Flag: ${tagYesNo(t.ai_flag)}

Writing sample:
${e.writing_sample || ''}
`;
  });
  return `Student: ${studentName}
IEP notes: ${iep}

I am sharing ${n} writing samples from this student in chronological order. Please analyze their growth as a writer.

${blocks.join('\n')}

Analyze the growth across all ${n} entries and return the JSON object.`;
}

function normalizeComparison(raw) {
  if (!raw || typeof raw !== 'object') throw new Error('Invalid comparison object');
  const c = { ...raw };
  if (typeof c.overall_growth_summary !== 'string') c.overall_growth_summary = String(c.overall_growth_summary ?? '');
  if (typeof c.what_stayed_strong !== 'string') c.what_stayed_strong = String(c.what_stayed_strong ?? '');
  if (!Array.isArray(c.growth_moments)) c.growth_moments = [];
  if (!Array.isArray(c.persistent_gaps)) c.persistent_gaps = [];
  if (!Array.isArray(c.entry_by_entry)) c.entry_by_entry = [];
  if (!c.next_instructional_step || typeof c.next_instructional_step !== 'object') {
    c.next_instructional_step = {
      move: '',
      rationale: '',
      try_this: '',
      iep_connection: null,
    };
  }
  if (typeof c.conference_script !== 'string') c.conference_script = String(c.conference_script ?? '');
  return c;
}

async function runComparison(userMessage) {
  const provider = resolveProvider();
  const { text, modelUsed, provider: providerUsed } = await provider.complete({
    system: SYSTEM_PROMPT,
    user: userMessage,
    kind: 'compare',
  });
  let raw;
  try {
    raw = parseModelJson(text);
  } catch (e) {
    console.error('[compare-progress] JSON parse failed. Raw model text (truncated):', text.slice(0, 4000));
    throw new Error(`Could not parse comparison response: ${e.message}`);
  }
  return { comparison: normalizeComparison(raw), modelUsed, providerUsed };
}

router.post('/', async (req, res) => {
  const body = req.body || {};
  const studentId = Number(body.student_id);
  const studentName = body.student_name != null ? String(body.student_name).trim() : '';
  const iepFlags = body.iep_flags;
  const entries = Array.isArray(body.entries) ? body.entries : [];

  if (!Number.isInteger(studentId)) return fail(res, 'student_id is required', 400);
  if (!studentName) return fail(res, 'student_name is required', 400);
  if (entries.length < 2) return fail(res, 'At least 2 entries are required', 400);
  if (entries.length > 6) return fail(res, 'Maximum 6 entries per comparison', 400);

  for (let i = 0; i < entries.length; i++) {
    const e = entries[i];
    if (!e || typeof e !== 'object') return fail(res, `Invalid entry at index ${i}`, 400);
    const eid = Number(e.entry_id);
    if (!Number.isInteger(eid)) return fail(res, `Invalid entry_id at index ${i}`, 400);
    if (e.writing_sample == null || String(e.writing_sample).trim().length < 1) {
      return fail(res, `Entry ${eid} is missing writing_sample`, 400);
    }
  }

  const entryIds = entries.map((e) => Number(e.entry_id));
  const unique = new Set(entryIds);
  if (unique.size !== entryIds.length) return fail(res, 'Duplicate entry_id in request', 400);

  const verify = await db.query(
    `SELECT id, date_submitted FROM entries WHERE student_id = $1 AND id = ANY($2::int[])`,
    [studentId, entryIds]
  );
  if (verify.rows.length !== entryIds.length) {
    return fail(res, 'One or more entries do not belong to this student', 400);
  }

  const byId = new Map(verify.rows.map((r) => [r.id, r]));
  const dayKey = (d) => {
    if (d == null) return '';
    if (d instanceof Date) return d.toISOString().slice(0, 10);
    return String(d).slice(0, 10);
  };
  let prevKey = null;
  let prevId = null;
  for (const id of entryIds) {
    const row = byId.get(id);
    const key = dayKey(row.date_submitted);
    if (prevKey != null) {
      if (key < prevKey) {
        return fail(res, 'entries must be in chronological order (oldest first)', 400);
      }
      if (key === prevKey && id <= prevId) {
        return fail(res, 'entries must be in chronological order (oldest first)', 400);
      }
    }
    prevKey = key;
    prevId = id;
  }

  const userMessage = buildUserMessage(studentName, iepFlags, entries);

  try {
    const { comparison, modelUsed, providerUsed } = await runComparison(userMessage);
    return ok(res, {
      comparison,
      student_id: studentId,
      entry_ids_compared: entryIds,
      generated_at: new Date().toISOString(),
      model_used: modelUsed,
      provider_used: providerUsed,
    });
  } catch (e) {
    const rawStatus = e.statusCode ?? (typeof e.status === 'number' ? e.status : null);
    const status =
      rawStatus === 503 || rawStatus === 401 || rawStatus === 429
        ? rawStatus
        : rawStatus >= 400 && rawStatus < 600
          ? rawStatus
          : 502;
    const msg = e.message || 'Comparison failed';
    if (status === 503) return fail(res, msg, 503);
    return fail(res, msg, status);
  }
});

module.exports = router;

File: desktop/src-tauri/src/routes/compare_progress.rs
//! Port of routes/compareProgress.js — AI growth comparison across 2-6
//! chronologically-ordered entries for one student.

use axum::extract::{FromRef, State};
use axum::http::StatusCode;
use axum::response::Response;
use serde_json::{json, Value};

use crate::ai;
use crate::ai::json::parse_model_json;
use crate::db::{blocking, Pool};
use crate::error::{ok_json, ApiError};
use crate::server::AppState;
use crate::settings;

const SYSTEM_PROMPT: &str = include_str!("../../../../prompts/compare-progress.system.txt");

fn tag_yes_no(tags: &Value, key: &str) -> &'static str {
    if tags[key] == Value::Bool(true) { "yes" } else { "no" }
}

fn build_user_message(student_name: &str, iep_flags: &Value, entries: &[Value]) -> String {
    let iep = iep_flags
        .as_str()
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .unwrap_or("None provided");
    let n = entries.len();

    let blocks: Vec<String> = entries
        .iter()
        .enumerate()
        .map(|(i, e)| {
            let t = &e["tags"];
            format!(
                "--- Entry {num}: {name} ({date}) ---\nWord count: {wc}\nTags: Claim: {claim}, Evidence: {ev}, Explanation: {ex}, Source Named: {sn}, Incomplete: {inc}, AI Flag: {ai}\n\nWriting sample:\n{sample}\n",
                num = i + 1,
                name = e["assignment_name"].as_str().unwrap_or(""),
                date = e["date"].as_str().unwrap_or(""),
                wc = e["word_count"].as_i64().unwrap_or(0),
                claim = tag_yes_no(t, "claim_present"),
                ev = tag_yes_no(t, "evidence_cited"),
                ex = tag_yes_no(t, "explanation_present"),
                sn = tag_yes_no(t, "source_named"),
                inc = tag_yes_no(t, "response_incomplete"),
                ai = tag_yes_no(t, "ai_flag"),
                sample = e["writing_sample"].as_str().unwrap_or(""),
            )
        })
        .collect();

    format!(
        "Student: {student_name}\nIEP notes: {iep}\n\nI am sharing {n} writing samples from this student in chronological order. Please analyze their growth as a writer.\n\n{}\n\nAnalyze the growth across all {n} entries and return the JSON object.",
        blocks.join("\n")
    )
}

fn normalize_comparison(mut raw: Value) -> Result<Value, String> {
    if !raw.is_object() {
        return Err("Invalid comparison object".to_string());
    }
    let obj = raw.as_object_mut().unwrap();
    for key in ["overall_growth_summary", "what_stayed_strong", "conference_script"] {
        let v = obj.get(key).cloned().unwrap_or(Value::Null);
        if !v.is_string() {
            let s = match v {
                Value::Null => String::new(),
                other => other.to_string(),
            };
            obj.insert(key.to_string(), Value::String(s));
        }
    }
    for key in ["growth_moments", "persistent_gaps", "entry_by_entry"] {
        if !obj.get(key).map(Value::is_array).unwrap_or(false) {
            obj.insert(key.to_string(), json!([]));
        }
    }
    if !obj.get("next_instructional_step").map(Value::is_object).unwrap_or(false) {
        obj.insert(
            "next_instructional_step".to_string(),
            json!({ "move": "", "rationale": "", "try_this": "", "iep_connection": null }),
        );
    }
    Ok(raw)
}

pub async fn compare(
    State(state): State<AppState>,
    axum::Json(body): axum::Json<Value>,
) -> Result<Response, ApiError> {
    let Some(student_id) = body["student_id"].as_i64() else {
        return Err(ApiError::bad_request("student_id is required"));
    };
    let student_name = body["student_name"].as_str().unwrap_or("").trim().to_string();
    if student_name.is_empty() {
        return Err(ApiError::bad_request("student_name is required"));
    }
    let entries = body["entries"].as_array().cloned().unwrap_or_default();
    if entries.len() < 2 {
        return Err(ApiError::bad_request("At least 2 entries are required"));
    }
    if entries.len() > 6 {
        return Err(ApiError::bad_request("Maximum 6 entries per comparison"));
    }

    let mut entry_ids: Vec<i64> = Vec::with_capacity(entries.len());
    for (i, e) in entries.iter().enumerate() {
        if !e.is_object() {
            return Err(ApiError::bad_request(format!("Invalid entry at index {i}")));
        }
        let Some(eid) = e["entry_id"].as_i64() else {
            return Err(ApiError::bad_request(format!("Invalid entry_id at index {i}")));
        };
        if e["writing_sample"].as_str().map(str::trim).unwrap_or("").is_empty() {
            return Err(ApiError::bad_request(format!("Entry {eid} is missing writing_sample")));
        }
        entry_ids.push(eid);
    }
    {
        let unique: std::collections::HashSet<i64> = entry_ids.iter().copied().collect();
        if unique.len() != entry_ids.len() {
            return Err(ApiError::bad_request("Duplicate entry_id in request"));
        }
    }

    // Ownership + chronological-order verification against the active DB.
    let pool: Pool = Pool::from_ref(&state);
    let ids_for_check = entry_ids.clone();
    let generated_at = blocking(&pool, move |conn| {
        let mut stmt = conn.prepare("SELECT id, date_submitted FROM entries WHERE student_id = ?")?;
        let owned: std::collections::HashMap<i64, String> = stmt
            .query_map([student_id], |r| Ok((r.get::<_, i64>(0)?, r.get::<_, String>(1)?)))?
            .collect::<Result<_, _>>()?;

        if !ids_for_check.iter().all(|id| owned.contains_key(id)) {
            return Err(ApiError::bad_request("One or more entries do not belong to this student"));
        }

        let mut prev: Option<(String, i64)> = None;
        for id in &ids_for_check {
            let key: String = owned[id].chars().take(10).collect();
            if let Some((prev_key, prev_id)) = &prev {
                if key < *prev_key || (key == *prev_key && *id <= *prev_id) {
                    return Err(ApiError::bad_request(
                        "entries must be in chronological order (oldest first)",
                    ));
                }
            }
            prev = Some((key, *id));
        }

        let now: String =
            conn.query_row("SELECT STRFTIME('%Y-%m-%dT%H:%M:%fZ', 'now')", [], |r| r.get(0))?;
        Ok(now)
    })
    .await?;

    let user_message = build_user_message(&student_name, &body["iep_flags"], &entries);

    let config = settings::load(&state.settings_path);
    let result =
        ai::complete(&state.http, &config, SYSTEM_PROMPT, &user_message, ai::Kind::Compare).await?;

    let comparison = parse_model_json(&result.text)
        .map_err(|e| {
            eprintln!(
                "[compare-progress] JSON parse failed. Raw model text (truncated): {}",
                &result.text.chars().take(4000).collect::<String>()
            );
            ApiError::new(StatusCode::BAD_GATEWAY, format!("Could not parse comparison response: {e}"))
        })
        .and_then(|raw| {
            normalize_comparison(raw).map_err(|e| ApiError::new(StatusCode::BAD_GATEWAY, e))
        })?;

    Ok(ok_json(json!({
        "comparison": comparison,
        "student_id": student_id,
        "entry_ids_compared": entry_ids,
        "generated_at": generated_at,
        "model_used": result.model,
        "provider_used": result.provider,
    })))
}
